### ISSUE TITLE
BSD support

### ISSUE THREAD

**Issue by andrei-pavel** [sysrepo/sysrepo#2910]
Been trying to check if sysrepo works on FreeBSD 13. v1 of libyang and sysrepo build succesfully. After solving `/dev/shm` not existing which is inherent to BSD, I then get an `[ERR]: Invalid permissions 17777777` and a `Failed to connect` on a `sysrepoctl -l` command. I don't provide those permissions explicitly via `-p` since they are not even allowed when listing YANG modules. And I don't know what file has those permissions if any. It also doesn't look like valid file permissions. I tried to debug it with ktrace and it seems to happen right after reading from `/etc/groups`.

After that, I decided to try the master branch, but I cannot get that to build, even after installing additional packages and working around CMake not being able to find various dependency headers. Same build process works on various Linux distributions.

Do you have plans of supporting BSD distributions? If so, I can provide details on the above. If not, this issue can be closed as far as I'm concerned.

Thank you!

**Comment by michalvasko** [comment]
Yes, we would like to support BSD as long as there are no major issues. I will test it in a virtual environment.

**Comment by michalvasko** [comment]
I have fixed some basic compilation problems but not sure what exactly to do about the shared memory path, how did you solve it?

**Comment by andrei-pavel** [comment]
Added `link /tmp/shm /dev/shm` to `/etc/devfs.conf` and restarted the machine.

I then required to `mkdir /tmp/shm` at each startup. Alternatively you could link it directly to `/tmp`, but you would have sysrepo data mixed with the rest of `/tmp` and that's rather undesirable.

A more complex configuration where `/dev/shm` has its own `tmpfs` drive is probably also possible, but I have not investigated.

I would have thought that `/dev/shm` needs something specific for device files different than `tmpfs`, but maybe not.

**Comment by andrei-pavel** [comment]
Sorry, after looking into the file, that's `link /tmp/shm shm`.

**Comment by andrei-pavel** [comment]
Oh, and it seems that `/tmp/shm` survived a reboot.

**Comment by michalvasko** [comment]
So I have played with it a bit and the current `devel` version of *sysrepo* should compile just fine. However, there seems to be some kind of problem with the mutexes when I get `invalid argument` error for the first SHM module lock. Not really sure what to do about that.

**Comment by razvan-becheriu** [comment]
I have successfully make sysrepo compile and work on macOS/BSD like system.
I hope I am not doing some extremely crazy actions by using a different path for shm and using access instead of eaccess system function as this will be described below:

After running ```cmake ..``` in ```build``` directory, I am applying the following patch:

```
diff --git a/src/common.c b/src/common.c
index feacff10..3cb62dac 100644
--- a/src/common.c
+++ b/src/common.c
@@ -67,6 +67,18 @@ const struct srplg_ntf_s *sr_internal_ntf_plugins[] = {
     &srpntf_lyb,    /**< default */
 };
 
+#define PTHREAD_MUTEX_ROBUST -1
+
+int pthread_mutex_consistent(pthread_mutex_t *mutex)
+{
+    return 0;
+}
+
+int pthread_mutexattr_setrobust(pthread_mutexattr_t *attr, int robustness)
+{
+    return 0;
+}
+
 /**
  * @brief Default module DS plugins.
  */
diff --git a/src/plugins/ds_lyb.c b/src/plugins/ds_lyb.c
index 5a6bddd4..41282d32 100644
--- a/src/plugins/ds_lyb.c
+++ b/src/plugins/ds_lyb.c
@@ -25,7 +25,6 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
-#include <sys/inotify.h>
 #include <sys/stat.h>
 #include <sys/types.h>
 #include <time.h>
``` 

Then I change the SR_SHM_DIR path in ```build/config.h```:

e.g.:

```
#define SR_SHM_DIR "/path/to/sysrepo/shm"
```

I prefer to use a persistent location so that I don't have to recreate the folder every time (hence the sysrepo git folder is used in the path)

Make sure you create the folder before you run any sysrepo bin command.

Also add a define for ```eaccess``` to ```access``` in ```build/common_lyb.h``` as access is not supported on macOS (it is recognized by make but not properly handled).

```
#define eaccess access
```

After making install, paths needs to be updated:
```
install_name_tool -change @rpath/libyang.2.dylib /opt/libyang/lib/libyang.2.dylib /opt/libyang/bin/yanglint
install_name_tool -change @rpath/libyang.2.dylib /opt/libyang/lib/libyang.2.dylib /opt/libyang/bin/yangre

install_name_tool -change @rpath/libyang.2.dylib /opt/libyang/lib/libyang.2.dylib /opt/sysrepo/bin/sysrepoctl
install_name_tool -change @rpath/libyang.2.dylib /opt/libyang/lib/libyang.2.dylib /opt/sysrepo/bin/sysrepocfg
install_name_tool -change @rpath/libyang.2.dylib /opt/libyang/lib/libyang.2.dylib /opt/sysrepo/bin/sysrepo-plugind

install_name_tool -change @rpath/libsysrepo.7.dylib /opt/sysrepo/lib/libsysrepo.7.dylib /opt/sysrepo/bin/sysrepoctl
install_name_tool -change @rpath/libsysrepo.7.dylib /opt/sysrepo/lib/libsysrepo.7.dylib /opt/sysrepo/bin/sysrepocfg
install_name_tool -change @rpath/libsysrepo.7.dylib /opt/sysrepo/lib/libsysrepo.7.dylib /opt/sysrepo/bin/sysrepo-plugind
```
I use ```/opt``` to install all libyang/sysrepo libraries, but I guess this is required is using system paths also, just make sure you use proper paths.

Use ```otool -L /opt/sysrepo/bin/sysrepoctl``` to see dependencies on macOS.

**Comment by razvan-becheriu** [comment]
If you have permissions errors on sysrepoctl just run:

```
chmod 664 repository/data/ietf*
chmod 664 repository/data/kea*
chmod 664 repository/data/sysrepo*
```

You might need to do this after you install new modules as well.

**Comment by jktjkt** [comment]
A lot of these patches are not required:

- the build system already checks for `eaccess()`, and will fall back to `access()` if the former is missing,
- there's a `SHM_DIR` CMake option which you can use, so there's no need to overwrite `config,h`.
 
However, the real blocker are robust mutexes. If your platform doesn't support them and you patch them out, you will see either deadlocks or silent data corruption in case any process which uses sysrepo crashes/dies.

**Comment by michalvasko** [comment]
@razvan-becheriu As @jktjkt said, try using the `devel` branch versions of the projects, there are some updates with regard to BSD support. But yes, not having robust mutexes makes *sysrepo* unable to recover from incorrect termination of applications.

**Comment by rcmcdonald91** [comment]
Any updates on this?

I started hacking on this today on FreeBSD main, I get a EINVAL from `pthread_mutex_timedlock` when running `sysrepoctl --list`. Does FreeBSD not support robust mutexes?

**Comment by rcmcdonald91** [comment]
https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=274396

**Comment by michalvasko** [comment]
Most of the changes are concerned with what seems like package information but you can provide the rest of the changes in a PR, with all the necessary checks or `#ifdef`s as needed.
